atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/CoordinationServiceRecorder.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.backup;
+
+import com.google.common.collect.Iterables;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.coordination.CoordinationService;
+import com.palantir.atlasdb.coordination.ValueAndBound;
+import com.palantir.atlasdb.internalschema.InternalSchemaMetadata;
+import com.palantir.atlasdb.internalschema.InternalSchemaMetadataState;
+import com.palantir.atlasdb.internalschema.persistence.CoordinationServices;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.impl.CheckAndSetResult;
+import com.palantir.atlasdb.timelock.api.CompletedBackup;
+import com.palantir.atlasdb.timelock.api.Namespace;
+import java.util.Optional;
+import java.util.function.Function;
+
+final class CoordinationServiceRecorder {
+    private final Function<Namespace, KeyValueService> keyValueServiceFactory;
+    private final Function<Namespace, Long> timestampSupplier;
+    private final SchemaMetadataPersister schemaMetadataPersister;
+
+    CoordinationServiceRecorder(
+            Function<Namespace, KeyValueService> keyValueServiceFactory,
+            Function<Namespace, Long> timestampSupplier,
+            SchemaMetadataPersister schemaMetadataPersister) {
+        this.keyValueServiceFactory = keyValueServiceFactory;
+        this.timestampSupplier = timestampSupplier;
+        this.schemaMetadataPersister = schemaMetadataPersister;
+    }
+
+    public void recordAtBackupTimestamp(Namespace namespace, long timestamp) {
+        fetchSchemaMetadata(namespace, timestamp)
+                .ifPresent(metadata -> schemaMetadataPersister.persistAtBackupTimestamp(namespace, metadata));
+    }
+
+    public boolean verifyFastForwardState(CompletedBackup completedBackup) {
+        Namespace namespace = completedBackup.getNamespace();
+        Optional<InternalSchemaMetadataState> fastForwardMetadata =
+                fetchSchemaMetadata(namespace, completedBackup.getBackupEndTimestamp());
+        schemaMetadataPersister.verifyFastForwardState(
+                namespace, fastForwardMetadata, timestampSupplier.apply(namespace));
```

Comment:
I'm pretty sure we never need a fresh timestamp here even if we did the old thing